crud/src/main/java/com/redhat/lightblue/mediator/Mediator.java

```diff
@@ -647,22 +651,12 @@ public Response explain(FindRequest req) {
         return response;        
     }
 
-    public static class BulkExecutionContext {
-        final Future<Response>[] futures;
-        final Response[] responses;
-
-        public BulkExecutionContext(int size) {
-            futures = new Future[size];
-            responses = new Response[size];
-        }
-    }
-
     protected void wait(BulkExecutionContext ctx) {
         for (int i = 0; i < ctx.futures.length; i++) {
             if (ctx.futures[i] != null) {
                 try {
                     LOGGER.debug("Waiting for a find request to complete");
-                    ctx.responses[i] = ctx.futures[i].get();
+                    ctx.setResponseAt(i, ctx.futures[i].get());
```

Comment:
> It would be interesting if we could allow responses to share an abstraction like a SizeCap or something of the sort. Then, we could have on SizeCap that all parallel responses shared, and then we wouldn't have to wait for them to each finish before summing their results (where it's likely some have already completed and we could already be well over the max by that time).

That's true. Unordered bulk requests are still dangerous, because all requests are executed in parallel and the limit is not effectively enforced. But nobody uses them afaik. For ordered bulk requests this should work reasonably well.

> Take this a step further ...
and use client side streaming :)

> As an aside, do we have any way to manage the size of the servlet-serving thread pool that the app server will use?

Yes we do, http executor's pool can be configured in threads subsystem of EAP. Joins are part of response, so they are included in the response size check. The problem I see is that we can introduce another bottleneck this way. At the very least this would require some tuning to get it right.